File: python/2d-dynamic-programming/burst-balloons.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        nums = [1] + nums + [1]
        cache = {}

        def dfs(left, right):
            if left < 1 or right < 1 or left > len(nums) - 2 or right > len(nums) - 2:
                return 0

            if (left, right) in cache:
                return cache[(left, right)]

            cache[(left, right)] = 0
            for i in range(left, right + 1):
                cache[(left, right)] = max(cache[(left, right)], nums[i] * nums[left - 1] * nums[right + 1] + dfs(left, i - 1) + dfs(i + 1, right))

            return cache[(left, right)]

        return dfs(1, len(nums) - 2)

File: c++/2d-dynamic-programming/burst-balloons.cc
class Solution {
public:
    int maxCoins(vector<int>& nums) {
        std::unordered_map<int, std::unordered_map<int, int>> cache;
        std::vector<int> balloons = {1};
        for (int num : nums) {
            balloons.push_back(num);
        }
        balloons.push_back(1);
        return burst_balloons(0, balloons.size() - 1, balloons, cache);
    }

private:
    int burst_balloons(int left, int right, std::vector<int>& nums, std::unordered_map<int, std::unordered_map<int, int>>& cache) {
        if (left == right) {
            return 0;
        }

        if (cache.contains(left) && cache[left].contains(right)) {
            return cache[left][right];
        }   

        cache[left][right] = 0;
        for (int i = left + 1; i < right; ++i) {
            cache[left][right] = std::max(cache[left][right], nums[left] * nums[right] * nums[i] + burst_balloons(left, i, nums, cache) + burst_balloons(i, right, nums, cache));
        }

        return cache[left][right];
    }
};
